### PR TITLE
IGNITE-21336 Handle ClosedChannelException gracefully

### PR DESCRIPTION
**Pull request by rpuch**
https://issues.apache.org/jira/browse/IGNITE-21336

Thank you for submitting the pull request.

To streamline the review process of the patch and ensure better code quality
we ask both an author and a reviewer to verify the following:

### The Review Checklist
- [ ] **Formal criteria:** TC status, codestyle, mandatory documentation. Also make sure to complete the following:  
\- There is a single JIRA ticket related to the pull request.  
\- The web-link to the pull request is attached to the JIRA ticket.  
\- The JIRA ticket has the Patch Available state.  
\- The description of the JIRA ticket explains WHAT was made, WHY and HOW.  
\- The pull request title is treated as the final commit message. The following pattern must be used: IGNITE-XXXX Change summary where XXXX - number of JIRA issue.
- [ ] **Design:** new code conforms with the design principles of the components it is added to.
- [ ] **Patch quality:** patch cannot be split into smaller pieces, its size must be reasonable.
- [ ] **Code quality:** code is clean and readable, necessary developer documentation is added if needed.
- [ ] **Tests code quality:** test set covers positive/negative scenarios, happy/edge cases. Tests are effective in terms of execution time and resources.

### Notes
- [Apache Ignite Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Java+Code+Style+Guide)

### CODE REVIEW

File: modules/network/src/main/java/org/apache/ignite/internal/network/netty/NettySender.java [hunk 1]
```diff
@@ -41,31 +48,131 @@ public class NettySender {
...
+                        recoverSendAfterChannelClosure(obj, currentChannel, triggerChannelRecreation);
+                    } catch (RuntimeException | AssertionError e) {
+                        LOG.error("An error while sending a message {}", e, obj.networkMessage());
+                    }
+                }
+            });
+        }
+    }
+
+    private static boolean completedSuccessfully(CompletableFuture<Void> writeFuture) {
```
Comment (sashapolo) [inline modules/network/src/main/java/org/apache/ignite/internal/network/netty/NettySender.java:134]: I'm pretty sure I've seen this method already, I suggest extracting it

File: modules/network/src/main/java/org/apache/ignite/internal/network/netty/NettySender.java [hunk 2]
```diff
@@ -41,31 +48,131 @@ public class NettySender {
...
+     * @return Future of the send operation (that gets completed when the message gets acknowledged by the receiver).
+     */
+    public CompletableFuture<Void> send(OutNetworkObject obj, Runnable triggerChannelRecreation) {
+        if (!obj.networkMessage().needAck()) {
+            // We don't care that the client might get an exception like ClosedChannelException or that the message
+            // will be lost if the channel is closed as it does not require to be acked.
+            return toCompletableFuture(channel.writeAndFlush(obj));
+        }
+
+        // Write in event loop to make sure that, if a ClosedSocketException happens, we recover from it without existing the event loop.
```
Comment (sashapolo) [inline modules/network/src/main/java/org/apache/ignite/internal/network/netty/NettySender.java:104]:
```suggestion
        // Write in event loop to make sure that, if a ClosedSocketException happens, we recover from it without exiting the event loop.
```

File: modules/network/src/main/java/org/apache/ignite/internal/network/netty/NettySender.java [hunk 3]
```diff
@@ -41,31 +48,131 @@ public class NettySender {
...
+        // concurrently.
+
+        Channel holderChannel = recoveryDescriptor.holderChannel();
+
+        if (holderChannel == null || holderChannel == currentChannel) {
+            // Our channel is being closed or is already completely closed, its pipeline might be (or not be) destroyed.
+            // But no new channel acquired the descriptor, so we are alone.
+
+            if (obj.shouldBeSavedForRecovery()) {
+                // It was not saved yet (as normally OutboundRecoveryHandler adds it to the unacknowledged messages queue, but
```
Comment (sashapolo) [inline modules/network/src/main/java/org/apache/ignite/internal/network/netty/NettySender.java:152]:
```suggestion
                // It was not saved yet (as normally OutboundRecoveryHandler adds it to the unacknowledged messages queue), but
```

File: modules/network/src/main/java/org/apache/ignite/network/DefaultMessagingService.java [hunk 4]
```diff
@@ -310,7 +311,14 @@ private CompletableFuture<Void> sendViaNetwork(
...
 
         return connectionManager.channel(consistentId, type, addr)
-                .thenComposeToCompletable(sender -> sender.send(new OutNetworkObject(message, descriptors)));
+                .thenComposeToCompletable(sender -> sender.send(
+                        new OutNetworkObject(message, descriptors),
+                        () -> triggerChannelCreation(consistentId, type, addr)
+                ));
+    }
+
+    private OrderingFuture<NettySender> triggerChannelCreation(@Nullable String consistentId, ChannelType type, InetSocketAddress addr) {
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/network/DefaultMessagingService.java:320]: To make it clear why `connectionManager.channel()` is used, to make the intention obvious.

File: modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryServerHandshakeManager.java [hunk 5]
```diff
@@ -98,6 +102,8 @@ public class RecoveryServerHandshakeManager implements HandshakeManager {
 
     private final BooleanSupplier stopping;
 
+    private final AtomicBoolean eventLoopSwitchPropagator = new AtomicBoolean();
```
[the remark this answers is not in the document]
Comment (rpuch) [reply modules/network/src/main/java/org/apache/ignite/internal/network/recovery/RecoveryServerHandshakeManager.java]: I thought a little bit more and just removed this manual HB propagation. Switching includes registration on the new event loop, and this includes execution of a task on the new event loop's thread; the future is completed on this new thread. This already established a HB.

**Review by sashapolo — APPROVED** [review]

**Pull request merged**